torch/_dynamo/variables/functions.py

```diff
@@ -248,10 +249,21 @@ def bind_args(self, parent, args, kwargs):
                         closure_cell_contents = AttrSource(
                             closure_cell, "cell_contents"
                         )
+
                         try:
-                            contents_var = VariableBuilder(
-                                parent, closure_cell_contents
-                            )(cell.cell_contents)
+                            if self.fn.__code__.co_filename == _module_dir(
+                                torch
+                            ) + "_functorch/autograd_function.py" and self.fn.__code__.co_freevars == (
+                                "original_forward",
+                                "original_setup_context",
+                            ):
+                                contents_var = SourcelessBuilder.create(
+                                    tx, cell.cell_contents
+                                )
+                            else:
+                                contents_var = VariableBuilder(
+                                    parent, closure_cell_contents
+                                )(cell.cell_contents)
```

Comment:
@zou3519 We have discussed this a bit and concluded with these comments:
https://github.com/pytorch/pytorch/blob/311af3b9889563e5441fee77bf432edc081f391d/torch/_dynamo/variables/misc.py#L666-L670

However, if we use autograd.Function inside of ```torch.vmap```, it seems we don't directly trace ```forward/backward``` function using ```AutogradFunctionApplyVariable```, but we trace ```_flat_vmap```:
https://github.com/pytorch/pytorch/blob/311af3b9889563e5441fee77bf432edc081f391d/torch/_functorch/vmap.py#L472-L480
This is a regular tracing, we can't control where to add guard like we did at ```AutogradFunctionApplyVariable```, so I use this sort of ugly way.

Reply:
Hmmm

Reply:
I don't understand this, I would expect that when we trace `_flat_vmap`, we still see an `AutogradFunctionApplyVariable`

Reply:
I think our design for this might be wrong. When we trace out the forward+backward, do we trace it on BatchedTensors?

Claim: We don't actually want to trace the autograd.Function.forward using BatchedTensor. Instead we should try to inline into the existing BatchedTensor x autograd.Function logic -- this constructs a second autograd.Function that operates on regular Tensors. That's the autograd.Function that should end up in the Dynamo-produced graph. All of the BatchedTensor x autograd.Function logic is in Python, so we should be able to inline through it.